### ISSUE TITLE
Registry on EC2 + S3 occasional Timeouts on docker push

### ISSUE THREAD

**Issue by jwhitcraft** [docker/distribution#707]
Hello,

I'm running v2.0.1 of the registry on a c4.2xl ubuntu ami with docker 1.7.0 and a few times a day when I do a docker push from another ec2 machine in the same VPC, i get a timeout error

```
Error pushing to registry: Head https://s3.amazonaws.com/my-docker-registry/docker/registry/v2/blobs/sha256/35/35fa464854b0bad3cb48d920d30159251ec3164067e07768eab876b62bd26d33/data?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=<left out>%2F20150717%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20150717T005149Z&X-Amz-Expires=1200&X-Amz-Signature=<left out>&X-Amz-SignedHeaders=host: read tcp 54.231.244.8:443: connection timed out
```

The last few lines of the registry log are

```
registry_1  | time="2015-07-17T00:56:31.335673369Z" level=info msg="s3.GetContent(\"/docker/registry/v2/repositories/test/ult-7.7.0.0beta2/_layers/sha256/35fa464854b0bad3cb48d920d30159251ec3164067e07768eab876b62bd26d33/link\")" trace.duration=550.107376ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).GetContent" trace.id=709dc503-b0d5-4bfa-a231-53ab24f4042c trace.line=55
registry_1  | time="2015-07-17T00:56:31.960985034Z" level=info msg="s3.Stat(\"/docker/registry/v2/blobs/sha256/35/35fa464854b0bad3cb48d920d30159251ec3164067e07768eab876b62bd26d33/data\")" trace.duration=625.210967ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=a03e7077-a9a8-4e56-ba98-fa2e9f415b05 trace.line=111
registry_1  | time="2015-07-17T00:56:32.597828653Z" level=info msg="s3.Stat(\"/docker/registry/v2/blobs/sha256/35/35fa464854b0bad3cb48d920d30159251ec3164067e07768eab876b62bd26d33/data\")" trace.duration=636.753926ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=7329e5c1-3555-4483-90d4-dd5fee5f660c trace.line=111
```

Here is the docker-compose.yml file that i used to start up the registry server

```yml
redis:
  ports:
  - 6379/tcp
  image: redis
registry:
  ports:
  - 443:5000/tcp
  environment:
    REGISTRY_HTTP_TLS_CERTIFICATE: /certs/my_cert.crt
    REGISTRY_HTTP_TLS_KEY: /certs/registry.test.net.key
    REGISTRY_STORAGE_CACHE_LAYERINFO: redis
    REGISTRY_REDIS_ADDR: redis:6379
    REGISTRY_STORAGE: s3
    REGISTRY_STORAGE_S3_ACCESSKEY: <omitted>
    REGISTRY_STORAGE_S3_BUCKET: my-docker-registry
    REGISTRY_STORAGE_S3_REGION: us-east-1
    REGISTRY_STORAGE_S3_SECRETKEY:  <omitted>
    REGISTRY_STORAGE_S3_V4AUTH: 'true'
    REGISTRY_STORAGE_S3_CHUNKSIZE: 5242880
  image: registry:2.0.1
  links:
    redis: redis
  volumes:
  - /etc/certs:/certs
```

**Comment by dmp42** [comment]
The way I read this is your ami can not communicate with S3 properly.
I would open a ticket with Amazon and mention intermittent network failures.

**Comment by jwhitcraft** [comment]
I did a test, and I spun up a new instance to run the registry on in a different zone and it still presented the same issues.

```
Error pushing to registry: Head https://s3.amazonaws.com/test-docker-registry/docker/registry/v2/blobs/sha256/a3/a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4/data?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=<omitted>%2F20150717%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20150717T160037Z&X-Amz-Expires=1200&X-Amz-Signature=<omitted>&X-Amz-SignedHeaders=host: read tcp 54.231.244.4:443: connection timed out
```

**Comment by dmp42** [comment]
@jwhitcraft unfortunately I can hardly think of anything else. `read tcp 54.231.244.4:443: connection timed out` does point to go failing to establish a connection to S3. Unless the go networking client is broken (which I doubt), this indicates a network issue communicating with S3.

We can, and maybe should, be more defensive / resistant in the code wrt such failures, but something is definitely wrong on AMZ side here.

**Comment by jwhitcraft** [comment]
So, while i see this sometimes with just doing a docker push. i'm seeing it almost every time when i use Jenkins to build my docker files and push to the registry, which is usually 4 builds a time. Could i just be over loading the registry with pushing that much?

**Comment by dmp42** [comment]
Interesting...

To make it super clear: the network error is between your docker daemon and S3, is that correct?

**Comment by jwhitcraft** [comment]
Yes, i do a `docker push <my registry>/test/something:latest` or jenkins does one for me, it will just and freeze up for like 40 minutes and then spit out the error from above of the error from #710

**Comment by dmp42** [comment]
Ok. So it's definitely a client code issue, and network issue between the docker daemon and S3. Registry is not involved here.

The fact that it freezes up 40 minutes is ridiculous - something is wrong in our code.

**Comment by stevvooe** [comment]
@jwhitcraft Are you running the registry in s3 or is s3 remote to your registry instance?

**Comment by stevvooe** [comment]
Perhaps, a fix for #432 would address this issue. It doesn't necessarily make sense to redirect in all cases if running a registry outside of s3, especially when data is not actually being fetched.

**Comment by jwhitcraft** [comment]
@stevvooe ,

The Registry code is running on an EC2 machine pointed at S3 for the storage driver, both are in the us-east-1 region.

**Comment by stevvooe** [comment]
@jwhitcraft Ok.

This sounds like we have two issues going on here:

1. Network failures in AWS to S3. You'll have to pursue this through AWS support.
1. Poor timeouts setting for layer existence checks. We'll need to pinpoint these by looking at the daemon logs. Make sure you have debug enabled by starting docker with `-D`.

**Comment by jwhitcraft** [comment]
@stevvooe,

Some info from AWS:

```
Thank you for your patience.
Usually if there is no data being send or received from a TCP socket, it will be closed.
A quick test showed that 20 seconds of inactivity will close it:

root@ubuntu:~# time telnet 54.231.244.8 443
Trying 54.231.244.8...
Connected to 54.231.244.8.
Escape character is '^]'.
Connection closed by foreign host.

real    0m20.432s
user    0m0.001s
sys     0m0.006s
```

I'm working with our network engineers on the VPC to S3 network issue, but it would seem like the S3 client should time out after 20 second of not seeing any data comeback since that is what S3 does.

I'm still working on getting a client setup to replicate this with debugging on .  We've switched to using file system storage till we can get the network figured out and I can give you more information.

**Comment by stevvooe** [comment]
@jwhitcraft Part of the issue here is that we haven't isolated the call that is hanging. I suspect we are seeing some sort of s3 timeout, but it may possibly be in the docker daemon client on redirect.

At this point, we'll take a peak at the reproduction and work from there. The daemon logs will help a lot here.

**Comment by jwhitcraft** [comment]
Since this bug was filed, we fixed the problem where were having with our VPC and S3, we had to add a [VPC Endpiont](https://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-endpoints.html) to allow our VPC direct access to S3 with out having to go over our via internal infrastructure which is in the west coast and we are using us-east-1.  

Once this was done, i am not seeing any problems and even the push and pulls are happening faster.  I've been running with debug mode one for my clients and I was unable to reproduce what was happening.

I would say that this issue can be closed since it was clearly a network issue.
